File: Algorithms/C++/00160-Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
        // check if last node is the same
        ListNode* tempA = headA;
        int lenA{0};
        while(tempA->next != nullptr){
            tempA = tempA->next;
            lenA++;
        }
        ListNode* tempB = headB;
        int lenB{0};
        while(tempB->next != nullptr){
            tempB = tempB->next;
            lenB++;
        }

        // if intersection
        if(tempA == tempB){
            int diff{0};
            if(lenA > lenB){
                diff = lenA - lenB;
                while(diff > 0){
                    headA = headA->next;
                    diff--;
                }
            } else if(lenB > lenA){
                diff = lenB - lenA;
                while(diff > 0){
                    headB = headB->next;
                    diff--;
                }
            }

            while(headA != nullptr){
                if(headA == headB) return headA;
                headA = headA->next;
                headB = headB->next;
            }

        }

        // if no intersection, return nullptr
        return nullptr;
    }
};

/*
    no hash set (with pointers)
    tc: O(n + m)
    sc: O(1)

    1st step:
        - we can figure out if two lists are intersected by checking the last node
    2nd step:
        - change length of lists such that they are equal
        - iterate thru both lists to find same node
        - return same node
        
    otherwise return nullptr if no intersection

    hash set (with ListNodes)

    - add list A into set
    - iterate through list B to see if ListNode is already contained in our set
    - if yes, return the ListNode that is contained
    - if no, return null

    tc: O(n + m)
    sc: O(n)



*/

File: Algorithms/Python3/00160-Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        # get length of A
        sizeA, sizeB = 0
        curr = headA
        while curr:
            sizeA += 1
            curr = curr.next
        # get length of B    
        curr = headB
        while curr:
            sizeB += 1
            curr = curr.next
        
        # balance A and B lists
        if sizeA > sizeB:
            for i in range(sizeA - sizeB):
                headA = headA.next
        elif sizeA < sizeB:
            for i in range(sizeB - sizeA):
                headB = headB.next
        
        # find intersected node
        while headA:
            if headA == headB:
                return headA
            headA = headA.next
            headB = headB.next

        return None

'''
    tc: O(n)
    sc: O(1)

    if we see a node in list1 == list2, then we know that there is an intersection
    [4,1,8,4,5]
    [5,6,1,8,4,5]
    return intersected node

    A: len = 5
    B: len = 6
        - move head of B by difference of len(B) - len(A)

    make lengths of A and B the same
    iterate through A and B to see if there is an interesected node
'''